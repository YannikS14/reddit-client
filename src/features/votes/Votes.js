import React, { useState } from 'react';
import Skeleton from 'react-loading-skeleton';
import 'react-loading-skeleton/dist/skeleton.css';
import Arrow from '../../assets/Arrow';
import { kFormatter } from '../post/Post';

export default function Votes({ votes, isLoading }) {
  const [vote, setVote] = useState('');
  const upvotes = kFormatter(votes);

  const handleClick = (targetBtn) => {
    vote === targetBtn ? setVote('') : setVote(targetBtn);
  };

  return (
    <div id="votes" className="pr-6 flex flex-col items-center">
      <button
        onClick={() => handleClick('upvote')}
        className="text-gray-500 dark:text-gray-300 hover:text-green-500"
        aria-label="upvote"
      >
        <Arrow
          className={`transform transition duration-300 ${
            vote === 'upvote' ? 'text-green-500 scale-125' : ''
          }`}
          testId="upvote-arrow"
        />
      </button>
      <p className="my-1.5 text-lg font-bold">
        {isLoading ? <Skeleton height={24} width={31} /> : upvotes}
      </p>
      <button
        onClick={() => handleClick('downvote')}
        className="text-gray-500 dark:text-gray-300 hover:text-red-500"
        aria-label="downvote"
      >
        <Arrow
          className={`transform transition duration-300 rotate-180 ${
            vote === 'downvote' ? 'text-red-500 scale-125' : ''
          }
    `}
          testId="downvote-arrow"
        />
      </button>
    </div>
  );
}
