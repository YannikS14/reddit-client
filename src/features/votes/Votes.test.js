import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '../test-utils';
import Votes from './Votes';

describe('Votes component', () => {
  it('toggles the upvote arrow classes after clicking the upvote button', async () => {
    render(<Votes votes="123" isLoading={false} />);

    expect(screen.getByTestId('upvote-arrow')).not.toHaveClass(
      'text-green-500 scale-125',
    );

    fireEvent.click(screen.getByLabelText('upvote'));
    expect(await screen.findByTestId('upvote-arrow')).toHaveClass(
      'text-green-500 scale-125',
    );

    fireEvent.click(screen.getByLabelText('upvote'));
    expect(await screen.findByTestId('upvote-arrow')).not.toHaveClass(
      'text-green-500 scale-125',
    );
  });

  it('toggles the downvote arrow classes after clicking the downvote button', async () => {
    render(<Votes votes="123" isLoading={false} />);

    expect(screen.getByTestId('downvote-arrow')).not.toHaveClass(
      'text-red-500 scale-125',
    );

    fireEvent.click(screen.getByLabelText('downvote'));
    expect(await screen.findByTestId('downvote-arrow')).toHaveClass(
      'text-red-500 scale-125',
    );

    fireEvent.click(screen.getByLabelText('downvote'));
    expect(
      await screen.findByTestId('downvote-arrow'),
    ).not.toHaveClass('text-red-500 scale-125');
  });

  it('matches the snapshot', () => {
    const tree = renderer
      .create(<Votes votes="123" isLoading={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
