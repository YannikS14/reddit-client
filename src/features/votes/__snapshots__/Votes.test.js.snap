// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Votes component matches the snapshot 1`] = `
<div
  className="pr-6 flex flex-col items-center"
  id="votes"
>
  <button
    aria-label="upvote"
    className="text-gray-500 dark:text-gray-300 hover:text-green-500"
    onClick={[Function]}
  >
    <svg
      className="h-6 w-6 pointer-events-none transform transition duration-300 "
      data-testid="upvote-arrow"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 11l3-3m0 0l3 3m-3-3v8m0-13a9 9 0 110 18 9 9 0 010-18z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </svg>
  </button>
  <p
    className="my-1.5 text-lg font-bold"
  >
    123
  </p>
  <button
    aria-label="downvote"
    className="text-gray-500 dark:text-gray-300 hover:text-red-500"
    onClick={[Function]}
  >
    <svg
      className="h-6 w-6 pointer-events-none transform transition duration-300 rotate-180 
    "
      data-testid="downvote-arrow"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 11l3-3m0 0l3 3m-3-3v8m0-13a9 9 0 110 18 9 9 0 010-18z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </svg>
  </button>
</div>
`;
